File: Java/D11/d11p3.java
/*
 * There are N number of small cartons and each carton will have a balloon of 
some color identified by a number. You are asked to select fixed number of 
P-cartoons as a subset every time and find the number of distinct number of 
colors of balloons in the number of cartoons "P" you pick everytime.

The resultant cartoons ans will be N-P+1 and return it.

The number of distinct number of balloons is:
    nums[i:i+p-1] = [nums[i], nums[i+1], ..., nums[i+p-1]].


Input Format:
-------------
Line-1: Two space separated integers, N and P
Line-2: N space separated integers, nums[]

Output Format:
--------------
Print the (N-P+1) integers as output.


Sample Input-1:
---------------
7 4
1 2 2 3 3 4 4

Sample Output-1:
----------------
3 2 3 2

Explanation: 
------------
The number of distinct elements in each subarray of size P goes as follows:
- nums[0:3] = [1,2,2,3] so ans[0] = 3
- nums[1:4] = [2,2,3,3] so ans[1] = 2
- nums[2:5] = [2,3,3,4] so ans[2] = 3
- nums[3:6] = [3,3,4,4] so ans[3] = 2


Sample Input-2:
---------------
6 3
1 1 1 1 1 1

Sample Output-2:
----------------
1 1 1 1

Explanation:
------------
The number of distinct elements in each subarray goes as follows:
- nums[0:2] = [1,1,1] so ans[0] = 1
- nums[1:3] = [1,1,1] so ans[1] = 1
- nums[2:4] = [1,1,1] so ans[2] = 1
- nums[3:5] = [1,1,1] so ans[3] = 1


Sample Input-3:
---------------
7 3
1 2 3 4 2 1 3

Sample Output-3:
----------------
3 3 3 3 3


*/

// using map

import java.util.*;

/**
 * d11p3
 */
public class d11p3 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int p = sc.nextInt();
        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }
        sc.close();
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < p; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
        }
        System.out.print(map.size() + " ");
        for (int i = p; i < n; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            map.put(nums[i - p], map.get(nums[i - p]) - 1);
            if (map.get(nums[i - p]) == 0) {
                map.remove(nums[i - p]);
            }
            System.out.print(map.size() + " ");
        }
        
    }
}

File: CPP/D11/d11p3.cpp
// A tenth standard student has been given a task, Given P number of subject marks 
// and a number I, He has to print the I-th least value of sums among all the 
// possible sub-arrays of marks.

// Input Format:
// -------------
// Line-1: Two space separated inetegers, P and I.
// Line-2: P space separated integers, marks[].

// Output Format:
// --------------
// Print the I-th least value of possible sums.



// Sample Input-1:
// --------------
// 3 4
// 3 2 4

// Sample output-1:
// ---------------
// 5

// Explanation: 
// ------------
// The subarrays of 3 2 4 are:
// 1st subarray: 3 the sum is 3
// 2nd subarray: 2 the sum is 2
// 3rd subarray: 4 the sum is 4
// 4th subarray: 3,2 the sum is 5
// 5th subarray: 2,4 the sum is 6
// 6th subarray: 3,2,4 the sum is 9

// The 4th smallest is 5


// Sample Input-2:
// ---------------
// 4 7
// 2 2 4 4

// Sample output-2:
// ----------------
// 8


// Explanation: 
// ------------
// The subarrays of 2 2 4 4 are

// 1st subarray: 2 the sum is 2
// 2nd subarray: 2 the sum is 2
// 3rd subarray: 4 the sum is 4
// 4th subarray: 4 the sum is 4
// 5th subarray: 2,2 the sum is 4
// 6th subarray: 2,4 the sum is 6
// 7th subarray: 4,4 the sum is 8
// 8th subarray: 2,2,4 the sum is 8
// 9th subarray: 2,4,4 the sum is 10
// 10th subarray: 2,2,4,4 the sum is 8

// The 7th smallest is 8


#include<bits/stdc++.h>
using namespace std;

int f(vector<int> nums, int k) {
    int count = 0;
    int sum = 0;
    int length = nums.size();
    int left = 0, right = 0;
    while (right < length) {
        sum += nums[right];
        while (sum > k) {
            sum -= nums[left];
            left++;
        }
        count += right - left + 1;
        right++;
    }
    return count;
}

int main(){
    int n,k;
    cin>>n>>k;
    vector<int> nums(n,0);
    int mini=nums[0],sum=0;
    for(int i=0;i<n;++i){
        cin>>nums[i];
        mini = min(mini,nums[i]);
        sum += nums[i];
    }
    int low = mini, high = sum;
    while (low < high) {
        int mid = (high - low) / 2 + low;
        int count = f(nums, mid);
        if (count < k)
            low = mid + 1;
        else
            high = mid;
    }
    cout<<low;
    return 0;
}